plugins/module_utils/cloud.py

```diff
@@ -1,221 +1,187 @@
-    Usage:
-        >>> backoff = _exponential_backoff()
-        >>> backoff
-        <function backoff_backoff at 0x7f0d939facf8>
-        >>> list(backoff())
-        [2, 4, 8, 16, 32, 60, 60, 60, 60, 60]
+        delay (int or float): initial delay.
+        backoff (int or float): backoff multiplier e.g. value of 2 will  double the delay each retry.
+        max_delay (int or None): maximum amount of time to wait between retries.
+        jitter (bool): if set to true, add jitter to the generate value.
     """
-    def backoff_gen():
-        for retry in range(0, retries):
-            sleep = delay * backoff ** retry
-            yield sleep if max_delay is None else min(sleep, max_delay)
-    return backoff_gen
-
 
-def _full_jitter_backoff(retries=10, delay=3, max_delay=60, _random=random):
-    """ Implements the "Full Jitter" backoff strategy described here
-    https://www.awsarchitectureblog.com/2015/03/backoff.html
-    Args:
-        retries (int): Maximum number of times to retry a request.
-        delay (float): Approximate number of seconds to sleep for the first
-            retry.
-        max_delay (int): The maximum number of seconds to sleep for any retry.
-            _random (random.Random or None): Makes this generator testable by
-            allowing developers to explicitly pass in the a seeded Random.
-    Returns:
-        Callable that returns a generator. This generator yields durations in
-        seconds to be used as delays for a full jitter backoff strategy.
-    Usage:
-        >>> backoff = _full_jitter_backoff(retries=5)
-        >>> backoff
-        <function backoff_backoff at 0x7f0d939facf8>
-        >>> list(backoff())
-        [3, 6, 5, 23, 38]
-        >>> list(backoff())
-        [2, 1, 6, 6, 31]
+    def __init__(self, delay, backoff, max_delay=None, jitter=False):
+        self.delay = delay
+        self.backoff = backoff
+        self.max_delay = max_delay
+        self.jitter = jitter
+
+    def __iter__(self):
+        self.current_delay = self.delay
+        return self
+
+    def __next__(self):
+        return_value = self.current_delay if self.max_delay is None else min(self.current_delay, self.max_delay)
+        if self.jitter:
+            return_value = random.uniform(0.0, return_value)
+        self.current_delay *= self.backoff
+        return return_value
+
+
+def _retry_func(func, sleep_time_generator, retries, catch_extra_error_codes, found_f, status_code_from_except_f):
+    counter = 0
+    for sleep_time in sleep_time_generator:
+        try:
+            return func()
+        except Exception as exc:
+            counter += 1
+            if counter == retries:
+                raise
+            status_code = status_code_from_except_f(exc)
+            if found_f(status_code, catch_extra_error_codes):
+                time.sleep(sleep_time)
+            else:
+                raise
+
+
+class CloudRetry:
```

Comment:
the module_utils_ec2 integration tests are failing with some impressive stack traces.

They assume that any exceptions passed to `status_code_from_exception` are an instance of class.base_class (see ec2.AWSRetry)

https://35a51eb02fb9ce5fc0f5-6787ea1c1013a96bfd0a81714124e020.ssl.cf1.rackcdn.com/488/b02aebad5940f5500e4ccb93c093bc00339b2426/check/ansible-test-cloud-integration-aws-py36_5/61951f5/job-output.txt